### PR TITLE
frontend/feat: #206 Added In-App Messaging

### PR DESCRIPTION
**Pull request by PraveenGongada**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Added In-App Messaging Feature

### CODE REVIEW

**Review by Vignesh-772 — CHANGES_REQUESTED** [review]

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/CommonJsInterface.java [hunk 1]
```diff
@@ -3135,6 +3136,54 @@ public void currentPosition(String str) {
...
+    }
+
+    @JavascriptInterface
+    public void storeCallBackMessageUpdated(final String channelId, final String uuid, final String callback){
+        ChatService.storeCallBackMessage = callback;
+        SharedPreferences sharedPref = context.getSharedPreferences(context.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = sharedPref.edit();
+                                editor.putString("CHAT_CHANNEL_ID", channelId);
+                                editor.apply();
+                                setKeysInSharedPrefs("CHAT_CHANNEL_ID", channelId);
```
[the remark this answers is not in the document]
Comment (PraveenGongada) [reply Frontend/android-native/app/src/main/java/in/juspay/mobility/CommonJsInterface.java:3293]: Done

File: Frontend/ui-customer/src/Screens/RideBookingFlow/HomeScreen/View.purs [hunk 2]
```diff
@@ -1568,7 +1589,7 @@ getQuotesPolling pollingId action retryAction count duration push state = do
 driverLocationTracking :: forall action. (action -> Effect Unit) -> (String -> action) -> (String -> action) -> (Int -> Int -> action) -> Number -> String -> HomeScreenState -> String-> Flow GlobalState Unit
 driverLocationTracking push action driverArrivedAction updateState duration trackingId state routeState = do
   _ <- pure $ printLog "trackDriverLocation2_function" trackingId
-  if ((isLocalStageOn RideAccepted) || (isLocalStageOn RideStarted)) && ((getValueToLocalStore TRACKING_ID) == trackingId) then do
+  if ((isLocalStageOn RideAccepted) || (isLocalStageOn RideStarted) || (isLocalStageOn ChatWithDriver)) && ((getValueToLocalStore TRACKING_ID) == trackingId) then do
```
Comment (Vignesh-772) [inline Frontend/ui-customer/src/Screens/RideBookingFlow/HomeScreen/View.purs:1592]: use `any`  function to handle this case

File: Frontend/ui-customer/src/Screens/RideBookingFlow/HomeScreen/View.purs [hunk 3]
```diff
@@ -1591,9 +1612,9 @@ driverLocationTracking push action driverArrivedAction updateState duration trac
           rideID = state.data.driverInfoCardState.rideId
           srcLat = (resp ^. _lat)
           srcLon = (resp ^. _lon)
-          dstLat = if state.props.currentStage == RideAccepted then state.data.driverInfoCardState.sourceLat else state.data.driverInfoCardState.destinationLat
-          dstLon = if state.props.currentStage == RideAccepted then state.data.driverInfoCardState.sourceLng else state.data.driverInfoCardState.destinationLng
-          markers = if (isLocalStageOn RideAccepted) then (driverTracking "" ) else (rideTracking "")
+          dstLat = if state.props.currentStage == RideAccepted || (isLocalStageOn ChatWithDriver) then state.data.driverInfoCardState.sourceLat else state.data.driverInfoCardState.destinationLat
```
Comment (Vignesh-772) [inline Frontend/ui-customer/src/Screens/RideBookingFlow/HomeScreen/View.purs:1615]: one variable is checked in state one and another from local storage why?

File: Frontend/ui-driver/src/Components/RideActionModal/View.purs [hunk 4]
```diff
@@ -244,14 +250,15 @@ endRide push config =
...
+      , padding (Padding 0 0 0 4)
       ] <> FontStyle.subHeading1 TypoGraphy
       )
   ]
 
 cancelRide :: forall w . (Action -> Effect Unit) -> Config -> PrestoDOM (Effect Unit) w
 cancelRide push config = 
   linearLayout 
-  [ width WRAP_CONTENT
+  [ width MATCH_PARENT
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Components/RideActionModal/View.purs:261]: why this is `match_parent` ?

File: Frontend/ui-driver/src/Screens/HomeScreen/View.purs [hunk 5]
```diff
@@ -608,7 +634,7 @@ enableCurrentLocation state = if (DA.any (_ == state.props.currentStage) [RideAc
 
 rideStatusPolling :: forall action. String -> Number -> HomeScreenState -> (action -> Effect Unit) -> (String -> action) -> Flow GlobalState Unit
 rideStatusPolling pollingId duration state push action = do 
-  if (getValueToLocalStore RIDE_STATUS_POLLING) == "True" && (getValueToLocalStore RIDE_STATUS_POLLING_ID) == pollingId && isLocalStageOn RideAccepted then do 
+  if (getValueToLocalStore RIDE_STATUS_POLLING) == "True" && (getValueToLocalStore RIDE_STATUS_POLLING_ID) == pollingId && (isLocalStageOn RideAccepted || isLocalStageOn ChatWithCustomer) then do
```
Comment (Vignesh-772) [inline Frontend/ui-driver/src/Screens/HomeScreen/View.purs:637]: use `any` function

**Review by Vignesh-772 — CHANGES_REQUESTED** [review]

**Comment by Vignesh-772** [comment]
Resolve conflicts

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/CommonJsInterface.java [hunk 6]
```diff
@@ -3278,6 +3272,59 @@ public void currentPosition(String str) {
...
+
+    @JavascriptInterface
+    public void startChatListenerService() {
+        try {
+            SharedPreferences sharedPref = context.getSharedPreferences(context.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+            String appState = sharedPref.getString("ACTIVITY_STATUS", "null");
+            Intent chatListenerService = new Intent(context, ChatService.class);
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && appState.equals("onPause"))
+            {
+                AlarmManager manager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
```
[the remark this answers is not in the document]
Comment (PraveenGongada) [reply Frontend/android-native/app/src/main/java/in/juspay/mobility/CommonJsInterface.java:3306]: To start service for devices with android version 12+

**Review by Vignesh-772 — APPROVED** [review]

**Comment by Vignesh-772** [comment]
Squash the commits

**Review by witcher-shailesh — APPROVED** [review]

**Pull request merged**